src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java

```diff
@@ -0,0 +1,555 @@
+     */
+    ClassEntry classEntry(Utf8Entry ne);
+
+    /**
+     * {@return A {@link ClassEntry} describing the class described by
+     * provided {@linkplain ClassDesc}}
+     * If a Class entry in the pool already describes this class,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param classDesc the symbolic descriptor for the class
+     */
+    default ClassEntry classEntry(ClassDesc classDesc) {
+        return classEntry(utf8Entry(Util.toInternalName(classDesc)));
+    }
+
+    /**
+     * {@return A {@link PackageEntry} describing the class whose internal name
+     * is encoded in the provided {@linkplain Utf8Entry}}
+     * If a Package entry in the pool already describes this class,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param nameEntry the constant pool entry describing the internal name of
+     *                  the package
+     */
+    PackageEntry packageEntry(Utf8Entry nameEntry);
+
+    /**
+     * {@return A {@link PackageEntry} describing the class described by
+     * provided {@linkplain PackageDesc}}
+     * If a Package entry in the pool already describes this class,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param packageDesc the symbolic descriptor for the class
+     */
+    default PackageEntry packageEntry(PackageDesc packageDesc) {
+        return packageEntry(utf8Entry(packageDesc.packageInternalName()));
+    }
+
+    /**
+     * {@return A {@link ModuleEntry} describing the module whose name
+     * is encoded in the provided {@linkplain Utf8Entry}}
+     * If a module entry in the pool already describes this class,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param moduleName the constant pool entry describing the module name
+     */
+    ModuleEntry moduleEntry(Utf8Entry moduleName);
+
+    /**
+     * {@return A {@link ModuleEntry} describing the module described by
+     * provided {@linkplain ModuleDesc}}
+     * If a module entry in the pool already describes this class,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param moduleDesc the symbolic descriptor for the class
+     */
+    default ModuleEntry moduleEntry(ModuleDesc moduleDesc) {
+        return moduleEntry(utf8Entry(moduleDesc.moduleName()));
+    }
+
+    /**
+     * {@return A {@link NameAndTypeEntry} describing the provided name and type}
+     * If a NameAndType entry in the pool already describes this name and type,
+     * it is returned; otherwise, a new entry is added and the new entry is
+     * returned.
+     *
+     * @param nameEntry the member name
+     * @param typeEntry the member field or method descriptor
+     */
+    NameAndTypeEntry natEntry(Utf8Entry nameEntry, Utf8Entry typeEntry);
```

Comment:
I would be inclined to rename more literally as `nameAndTypeEntry`, which is consistent with the naming convention in all other cases in this interface AFAICT. (FWIW i keep reading nat as "not a type"!)

Reply:
Good point, renamed.
Thanks.